### ISSUE TITLE
"WritableDirectory cannot view preexisting TTrees" when extending a uproot-created TTree

### ISSUE THREAD

**Issue by yipengsun** [scikit-hep/uproot5#606]
Hi,

I'm trying to merge some large ntuples (between 400 MB to 4 GB, approximately) with a small friend ntuple.

Since the friend tree is aligned with the main tree, there's no manual alignment required. I used `library='numpy'` to read branches chunk by chunk and extend the output tree:

```python
#!/usr/bin/env python

import uproot


def getTreeSpec(tree):
    raw = tree.typenames()
    return {k: v.replace("_t", "") for k, v in raw.items()}


stepSize = 2500
#  stepSize = 10000
addBrName = 'probe_UBDT'

mainInput = 'raw.root'
friendInput = 'friend.root'
outputRootFile = uproot.recreate('merged.root')

mainRootFile = uproot.open(mainInput)
trees = [t.replace(";1", "") for t in mainRootFile if "DecayTree" in t]

for t in trees:
    mainSpec = getTreeSpec(mainRootFile[t])
    mainSpec[addBrName] = 'float'
    outputRootFile.mktree(t, mainSpec)

    mainDF = uproot.iterate(
        f"{mainInput}:{t}", step_size=stepSize, library="np"
    )
    friendDF = uproot.iterate(
        f"{friendInput}:{t}",
        [addBrName],
        step_size=stepSize,
        library="np",
    )

    for m, f in zip(mainDF, friendDF):
        m[addBrName] = f[addBrName]
        outputRootFile[t].extend(m)
```

But I'd get the following error after a few chunks have been extended:

```
Traceback (most recent call last):
  File "./test.py", line 38, in <module>
    outputRootFile[t].extend(m)
  File "/home/syp/misc/researches/lhcb-data_analysis/MuonBDTPid/.virtualenv/lib/python3.8/site-packages/uproot/writing/writable.py", line 957, in __getitem__
    return self._get_del_search(where, True)
  File "/home/syp/misc/researches/lhcb-data_analysis/MuonBDTPid/.virtualenv/lib/python3.8/site-packages/uproot/writing/writable.py", line 916, in _get_del_search
    step = step[item]
  File "/home/syp/misc/researches/lhcb-data_analysis/MuonBDTPid/.virtualenv/lib/python3.8/site-packages/uproot/writing/writable.py", line 957, in __getitem__
    return self._get_del_search(where, True)
  File "/home/syp/misc/researches/lhcb-data_analysis/MuonBDTPid/.virtualenv/lib/python3.8/site-packages/uproot/writing/writable.py", line 950, in _get_del_search
    return self._get(item, cycle)
  File "/home/syp/misc/researches/lhcb-data_analysis/MuonBDTPid/.virtualenv/lib/python3.8/site-packages/uproot/writing/writable.py", line 987, in _get
    raise TypeError(
TypeError: WritableDirectory cannot view preexisting TTrees; open the file with uproot.open instead of uproot.recreate or uproot.update
```

And the output file `merged.root` would contain 25000 (`10*stepSize`) entries. 

If I bump `stepSize = 10000`, then for the sample file (~440 MB), the merge would be successful. But for larger files, I got the same error, and at that point the merged file contains 100000 (`10*stepSize`) entries.

### Sample files

The sample files can be (hopefully) accessed at [here](https://cernbox.cern.ch/index.php/s/d3NUexFAmvkjqtB)

### System info

- gcc: 10.3.0
- Python: 3.8.9
- `uproot`: 4.2.3
- `numpy`: 1.20.3

**Comment by YSelfTool** [comment]
I can confirm this bug. I think it happens only if the tree is in a (sub)directory, because when exceeding the number of baskets, `_move_tree` moves the tree in `WritableFile._trees`, but not in the (sub)directory.
Here's a simple reproducer:
```
#!/usr/bin/env python3

import uproot
import awkward as ak
import numpy as np

# works
with uproot.recreate("test.root") as root_file:
    root_file.mktree("tree", {"branch": np.float32}, initial_basket_capacity=3)
    for i in range(4):
        print(i)
        root_file["tree"].extend({"branch": np.zeros(10)})

# works, too
with uproot.recreate("test.root") as root_file:
    root_file.mktree("dir/tree", {"branch": np.float32}, initial_basket_capacity=3)
    for i in range(4):
        print(i)
        tree = next((tree for tree in root_file._file._trees.values() if "/".join(tree.path) == "dir/tree"))
        tree.extend({"branch": np.zeros(10)})

# doesn't work
with uproot.recreate("test.root") as root_file:
    root_file.mktree("dir/tree", {"branch": np.float32}, initial_basket_capacity=3)
    for i in range(4):
        print(i)
        root_file["dir/tree"].extend({"branch": np.zeros(10)})
```

**Comment by jpivarski** [comment]
This was a subtle issue with caches, solved by manually checking to see if any active (writable) TTrees own the TDirectory they live in, and if so, that should supersede any subdirectory owned by a root directory.

Anyway, the symptom was quite a corner case: it only happens if you're filling a TTree inside a subdirectory (as @YSelfTool said) and you pull the TTree from the TDirectory via

```python
outputRootFile[t].extend(...)
```

or

```python
root_file["dir/tree"].extend(...)
```

instead of holding a reference to it like

```python
writable_tree.extend(...)
```

_and also_ cross a threshold in which the WritableTree has to move itself in the file because its metadata changes size (which is why it depends on `stepSize`).

But I'm glad you found it! The fix is targeting both 5.0.0rc2 (`main`, #677) and 4.3.5 (`main-v4`, #678).
